### PR TITLE
Perplexity runner for LMs

### PR DESCRIPTION
**Pull request by jlibovicky**
This PR introduces a runner that computes perplexity for each sentence fed to a decoder. It is supposed to be used for language models.

### CODE REVIEW

File: neuralmonkey/runners/runner.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def __init__(self, all_coders, fetches, vocabulary, postprocess):
         self._postprocess = postprocess
 
         self.decoded_sentences = []
-        self.result = None  # type: Option[ExecutionResult]
+        self.result = None  # type: Optional[ExecutionResult]
```
Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: todle se posere pac to neni importovany

File: neuralmonkey/runners/perplexity_runner.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        """Get the feedables and tensors to run."""
+        return self._all_coders, {'xents': self._xent_op}, {}
+
+    def collect_results(self, results: List[Dict]) -> None:
+        perplexities = np.mean([2 ** res['xents'] for res in results], axis=0)
+        xent = float(np.mean([res['xents'] for res in results]))
+        self.result = ExecutionResult(
+            outputs=perplexities.tolist(),
+            losses=[xent],
+            scalar_summaries=None,
```
Comment (jindrahelcl) [inline neuralmonkey/runners/perplexity_runner.py]: nešly/neměly by se sem narvat nějaký sumarís?

File: tests/tests_run.sh [hunk 3]
```diff
@@ -12,6 +12,7 @@ bin/neuralmonkey-train tests/post-edit.ini
 bin/neuralmonkey-train tests/factored.ini
 bin/neuralmonkey-train tests/classifier.ini
 bin/neuralmonkey-train tests/labeler.ini
+bin/neuralmonkey-train tests/language-model.ini
```
Comment (jindrahelcl) [inline tests/tests_run.sh]: máš tady pomlčku a tam podtržítko

File: neuralmonkey/runners/perplexity_runner.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/runners/perplexity_runner.py]: Nejsou potřeba, crossentropy (jako loss) a perplexity (jako evaluator) se tam zapíšou v training loop.

Comment (jindrahelcl) [inline neuralmonkey/runners/perplexity_runner.py]: jo jo..

**Comment by jlibovicky** [comment]
Padají testy, protože output series je perplexity a ve validačních datech není žádná output series perplexity a tak to říká warning a protože testy jsou strict, tak to padá. Co s tím? Vypnout strict na language model nebo řešit nějak systematičtěji?

**Comment by jindrahelcl** [comment]
Představoval bych si, že ten language model k tý větě vypíše tu perplexitu.. Záleží jak se s tim chceš dělat, mě by ani nevadilo, kdyby se v tomhle případě vypnulo strict.

**Comment by jindrahelcl** [comment]
vypadá to, že vypínání strict módu nefunguje. @tomasmcz podívej se prosím do toho logu a na to, jak se pouští ten test

**Comment by tomasmcz** [comment]
`NEURALMONKEY_STRICT` zapíná striktní error když je neprázdná, musíš smazat tu nulu. Pokud si myslíte, že by se to mělo dát vypnout i nulou, založte issue, příští týden to předělám.

Nicméně ten warning by se měl vyřešit nějak jinak, v případě, kde nemá smysl uvádět referenci, by si to nemělo stěžovat, že žádnou nemá. K čemu je ten warning vůbec dobrý?

**Comment by jlibovicky** [comment]
Warning je tam k tomu, aby řekl uživateli, že dělá něco divně. Když víš, že to, co děláš divně, je to v pořádku Přečteš si to a řekneš si, že ti to neva. Model, co nemá výstup a jenom počítá perplexitu, je tak divný případ, že by se podle mě sneslo, kdyby to psalo warning.

Ale pokud si myslíte, že bude lepší to udělat tak, aby to newarningovalo, klidně to můžu udělat. V konfiguraci se nastaví, že reference evaluatoru bude `None`, spousta typů se změní na `Optional`, někde přibude pár ifů a bude to.

@tomasmcz: Nebo myslíš, že to zakládá nebezpečný precedent, že když ignorujeme warning jednou, budeme ignorovat i další, až dospějeme k tomu, že žádný warning už nikdy nebude k ničemu?

Ad vypínání strict módu: Bez ohledu na předchozí myslím, že by to mělo jít. Jako neznalý uživatel bych předpokládal `1` zapni, `0` vypni.

**Comment by jindrahelcl** [comment]
Co se týče toho zapínání strict módu - dokud to je jen boolean, dává mi smysl to vypínat prázdným stringem, protože to je The Unix-way how to do it.. Nuly jsou v unixu true: ``if [ 0 ] ; then echo "ahoj"; fi``

**Comment by tomasmcz** [comment]
Je nějaký způsob, jak se mi může omylem  stát, že nebudu mít referenci a nebude to fatalní? Předpokládal bych, že když budu psát překladač a tu referenci tam nedám, tak si bude nějaký error function stěžovat, že mu chybí data.

V jaké situaci je tenhle konkrétní warning užitečný?

**Comment by jlibovicky** [comment]
Vida, to je dobrý postřeh. Myslím, že to tím pádem spadne, až když to bude chtít něco feedovat při trénování.

Ten warning může být užitečný při volání run scriptu. Tam se mi klidně může stát, že referenci chci mít a zapomenu ji. Stejně tak dobře jí vůbec chtít mít nemusím.

**Comment by jlibovicky** [comment]
Ono to hlavně je tím, že ty waningy vznikly jako neformální upozornění, že něco nemusí být tak, jak má. Jakmile ty warningy formalizujeme tak, že zavedem strict mode, kde mají být chybami, tak to výrazně mění, co znamenají. Asi by si to chtělo vygrepovat, kde všude jsou a rozmyslet si jejich význam v nové situaci.

**Comment by tomasmcz** [comment]
A ještě jiná věc: podle mne by mělo být možné ke konfigu s warningem napsat ekvivalentní konfig bez něj. Například mám warning na nepoužitou sekci. Není to error, protože sekce navíc ničemu nevadí, ale indikuje to pravděpodobnou chybu na straně uživatele. Jako uživatel se na to podívám, zjistím, že je to něco co mi tam omylem zůstalo, tak to smažu. Nebo zjistím, že mám něco špatně a měl jsem na to někde odkazovat. Nebo je to něco, co si tam chci nechat, ale teď se to nepoužívá, tak to můžu zakomentovat. Každopádně dostanu konfig (v případě smazání nebo zakomentování ekvivalentní s tím původním), který je bez warningu.
Situace, kdy existuje rozumně použitelný konfigurák, který se nedá napsat tak, aby to neházelo warning, podle mne není dobře.

**Comment by tomasmcz** [comment]
Ani v tom run scriptu nechápu, proč by to měl být warning. Buď v něm počítám nějaké skóre vůči referenci, pak to při absenci reference padne. Nebo ji tam nepotřebuju. A pak se to bude lišit jenom tím, že nebude ve výpisu a tam nepotřebuju mít warning, abych viděl, že jí tam nemám. Možná by bylo dobré někam do logu napsat, že nemám referenci, tak ji nevypisuji, ale podle mne tam nepatří warning, zvláště v případě, že existují scénaře, kde je absence reference naprosto v pořádku.

**Comment by jlibovicky** [comment]
No pokud zadefinuješ warning, tak jak jsi ho zadefinoval (a já proti tomu nic nemám) tak to, co se dřív psalo jako warning (jakákoli zpráva, která říká jen o něco málo podezřelejší věc než "decoder initialized" nebo, že vocabulary rozlišuje 20 000 tokenů), warning nebude. A to je i tenhle případ.

**Comment by jlibovicky** [comment]
Tak takhle už bych to mergoval.

Warningy podobného druhu už v code base podle mě nejsou. Vizte `grep -r 'warn(' neuralmonkey/`

File: neuralmonkey/learning_utils.py [hunk 4]
```diff
@@ -454,15 +454,6 @@ def _print_examples(dataset: Dataset,
         output_series_names = [s for s in output_series_names
                                if s in val_preview_output_series]
 
-    if not target_series_names:
-        warn("No reference series to preview during validation")
-
-    if not source_series_names:
-        warn("No source series to preview during validation")
-
-    if not output_series_names:
```
Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: Tohle by tu bejt mělo, ne? A možná i jako error.

Nechal bych ty ostatní jako normální log, aby bylo hned jasný, že tam nejsou..

Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Buď tam source series má být a není, takže to muselo spadnout už předtím nebo tam nemá být vůbec (co já vím, třeba budeš chtít samplovat z ničeho) a pak tě to nepřekvapí, ne?

Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: mluvil jsem o output series.

Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: nicméně přijde mi užitečný, aby tam bylo napsaný, i když tam nebude source nebo target. Nemusí to bejt warning.

Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Tak to stejný, ne? Nemít output series = nemít runner a to by mělo spadnout už někde předtím.

Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: Takže by tam měl být `assert`..
